LaTeX misc

GinoGiotto opened the pull request:
Better analyzed commit-by-commit:

* Fix #4651.

  The issue is the following: after an error is displayed due to incorrect LaTeX, a `grep` command searches the word 'invalid', 
  which does not appear when errors are present. Since the verifier fails only when 'invalid' is found, most errors go undetected. I replaced 'invalid' with 'no output PDF file produced!', which does appear when the `\bm` error is displayed. I cannot be certain that 'no output PDF file produced!' appears for all errors, but I've done a few tests with the output produced by `latex_collector.py` and it always did. Notice that after this change, done in commit 870474e76d5aa20d47b87322d078c497a211ce0b, the CI fails as expected: https://github.com/GinoGiotto/set.mm/actions/runs/14932386410/job/41951665938.

* Replace all occurrences of `\bm` with `\pmb`. The latter does not require additional packages and therefore fixes the above failing check.

* Remove support for the `accents` package in the verifiers. This is in accordance with [this comment](https://github.com/metamath/set.mm/issues/4196#issuecomment-2655060831), which belongs to a discussion about attempting to make TeX installation faster. The accent package is used by two latexdefs only, and only for `\underaccent` (I provide an equivalent replacement below). Conventions encourage a minimalist approach for LaTeX packages:

  <pre>
  When writing LaTeX definitions, contributors should favor simplicity over perfection of the display,
  and should only use core LaTeX symbols or symbols from standard packages; if packages other than
  amssymb, amsmath, mathtools, mathrsfs, phonetic, graphicx are needed, this should be discussed.
  </pre>

* Remove support for the `accents` package in `latex_collector.py`.  As a result, the LaTeX verifier correctly complains that `\underaccent` is not recognised: https://github.com/GinoGiotto/set.mm/actions/runs/14933046796/job/41953899527 (I'm mainly doing this to be confindent that it does actually fail when it is supposed to).

* Fix latexdefs that use `\underaccent`. The alternative `\d` accomplishes the exact same thing (displays a dot under a symbol or text) and it does not require additional packages. The previous failing check is fixed as a result: https://github.com/GinoGiotto/set.mm/actions/runs/14933229329.

* Conventions: remove `graphicx` from the list of accepted packages. This package used to serve the sole purpose of providing a feature to "manually" rotate iota. Such functionality has not been used since the LaTeX changes of 2023, when the rotated iota was replaced by `\riota`. The `graphicx` package has already been removed from metamath.exe (point 3 of https://github.com/metamath/metamath-exe/pull/133) and it's not supported by the verifiers and by `latex_collector.py`, so it's better to align the conventions with the current state.

* Add LaTeX generation checks for iset.mm and nf.mm. Since I put a lot of effort to fix iset.mm and nf.mm latexdefs as well, I would like those to stay functional too. The verification time for both is negligible. As a result, the verifiers correctly complain about `\underaccent` in iset.mm https://github.com/GinoGiotto/set.mm/actions/runs/14933603330/job/41955726546.

* Fix `\underaccent` in iset.mm and nf.mm, as done in set.mm.

Unfortunately, I did not find a way to represent a rotated iota without requiring an additional package. So I guess we will have to stick with phonetic.sty for now. ¯\\_(ツ)_/¯

GinoGiotto commented:
Actually 'invalid' can appear when warnings are present https://github.com/metamath/set.mm/pull/3100#issuecomment-1475359506, so I'll add it back together with 'no output PDF file produced!", since we probably want to detect warnings too.

wlammen commented:
I tried to follow the instructions:
I have a notebook with a native Linux installed (a variant of Ubuntu 24.04 LTS).  I do not use Latex on my system, but for this test I installed:

```
sudo apt-get install texlive-latex-base
sudo apt-get install texlive-fonts-recommended
sudo apt-get install texlive-fonts-extra

```
after which a
`which pdflatex`
executed successfully.
This was all to run the bash commands described in
https://github.com/metamath/set.mm/issues/4651
Unfortunately, the result was still an error:

> ! LaTeX Error: File `accents.sty' not found.

I looked for this file and found it on [accents.sty](https://github.com/chiamingyen/extensions/blob/master/extensions/portableLatex/MiKTeX/texmf/tex/latex/bezos/accents.sty).
This is a script file, so copying the contents of the page, and creating a file from it is not difficult.
The instructions given here [installing](https://tex.stackexchange.com/questions/73016/how-do-i-install-an-individual-package-on-a-linux-system) did not help me, neither the site [ctan](https://ctan.org/tex-archive/macros/latex/contrib/accents).
The problem is, I need more time than I have to find out what exactly is missing, or where I fail.

Sorry.

Edit
OK I found the install sequence in the updated verifiers.yml file.  My system now reflects this state.  So now I need the updated set.mm and iset.mm file.
Downloaded your branch from github.  So everything should be fine for a test now: The system has exactly the capabilities for a Latex generation, as requested by this pull request, and the files matches those you changed.

When executing the yml script by hand, the following command fails:
`python3 scripts/latex_collector.py set.mm list-set.tex`
resulting in 

> /home/wolf/metamath/test/set.mm-mul8_a/scripts/latex_collector.py:39: SyntaxWarning: invalid escape sequence '\l'
  output.write("\\newsavebox\ltmcbox\n")

Should this be a 

output.write("\\\\newsavebox\\\\ltmcbox\n")

instead?  I fixed this in my version, and then the script seems to run fine.

GinoGiotto commented:
> When executing the yml script by hand, the following command fails:

Good catch, I fixed this in 096101d. I also added a few other words to look for when errors appear. They are probably redundant, but better to be safe than sorry.

> Unfortunately, the result was still an error:
 ! LaTeX Error: File `accents.sty' not found.

Ideally metamath.exe would follow suit. I made a PR https://github.com/metamath/metamath-exe/pull/182 to align it to these changes. In practice it's not strictly needed that the two proposals are merged at the same time, but it would be the preferred situation.

wlammen commented:
Looking here [boxes](https://tex.stackexchange.com/questions/288399/reuse-of-newsavebox-and-avoiding-compilation-warnings), the syntax should be
`\newsavebox{\ltmcbox}`
The python3 script generates a file _list-set.tex_, and when I open it in a capable editor, syntax highlighting does not cover the currently created line.
For python, the \ need to be escaped as duplicated \\\\

Running
`pdflatex -halt-on-error list-set.tex 2>&1 | tee ,latex-output`

yields
...
> kpathsea: Running mktextfm rsfs10
mktextfm: Running mf-nowin -progname=mf \mode:=ljfour; mag:=1; ; nonstopmode; input rsfs10
This is METAFONT, Version 2.71828182 (TeX Live 2023/Debian) (preloaded base=mf)

> kpathsea: Running mktexmf rsfs10

> ! I can't find file `rsfs10'.
<*> ...ljfour; mag:=1; ; nonstopmode; input rsfs10
                                                  
> Please type another input file name
! Emergency stop.
<*> ...ljfour; mag:=1; ; nonstopmode; input rsfs10
                                                  
> Transcript written on mfput.log.
grep: rsfs10.log: Datei oder Verzeichnis nicht gefunden
mktextfm: `mf-nowin -progname=mf \mode:=ljfour; mag:=1; ; nonstopmode; input rsfs10' failed to make rsfs10.tfm.
kpathsea: Appending font creation commands to missfont.log.

> ! Font U/rsfs/m/n/10=rsfs10 at 10.0pt not loadable: Metric (TFM) file not found
.
<to be read again>
 
> relax 
l.16 \verb!ascii! & \LaTeX
                          \ \\
!  ==> Fatal error occurred, no output PDF file produced!
Transcript written on list-set.log.

The missing file can be found here [ctan](https://ctan.org/tex-archive/fonts/rsfs).  But it is not made automatically available by  the instructions in the verifyer.yml (as I understand them).

GinoGiotto commented:
> Looking here [boxes](https://tex.stackexchange.com/questions/288399/reuse-of-newsavebox-and-avoiding-compilation-warnings), the syntax should be
\newsavebox{\ltmcbox}
The python3 script generates a file list-set.tex, and when I open it in a capable editor, syntax highlighting does not cover the currently created line.

I'm using TeXstudio. The syntax highlighting looks fine to me:

![4](https://github.com/user-attachments/assets/a570399b-09ce-452b-a6d9-888be66c2dda)

If we apply your proposal, the color doesn't change:

![2](https://github.com/user-attachments/assets/3a8e79b9-a7b9-4f68-9f4a-b3c71b3403a1)

The lack of parenthesis doesn't cause warnings or errors, so I don't see where the issue is. If it's preferred anyway, I changed it: 65db599

>Running
>pdflatex -halt-on-error list-set.tex 2>&1 | tee ,latex-output
>yields
> ...
> The missing file can be found here [ctan](https://ctan.org/tex-archive/fonts/rsfs). But it is not made automatically available by the instructions in the verifyer.yml (as I understand them).

The [rsfs](https://ctan.org/tex-archive/fonts/rsfs) fonts are part of the [calrsfs](https://ctan.org/pkg/calrsfs) package, which is included in [texlive-latex-extra](https://packages.debian.org/sid/texlive-latex-extra). So I think the issue is that you did not install [texlive-latex-extra](https://packages.debian.org/sid/texlive-latex-extra). Notice that the verifiers do not complain about this, and your final error message `==> Fatal error occurred, no output PDF file produced!` would have caused them to fail.

wlammen commented:
I uninstalled everything texlive related and reinstalled the packages according to the list in veryfyers.yml.  This time the script finished with success, but still issues following warnings:
...
> (/usr/share/texlive/texmf-dist/tex/latex/amsfonts/ueuf.fd)

> Package longtable Warning: Column widths have changed
(longtable)                in table 1 on input line 1786.

>[1{/var/lib/texmf/fonts/map/pdftex/updmap/pdftex.map}] [2] [3] [4] [5] [6]
[7] [8] [9] [10] [11] [12] [13] [14]

> Package longtable Warning: Table widths have changed. Rerun LaTeX.

> [15] (./list-set.aux)

...
Similar with iset.mm, nf.mm.

Except for these warnings, I can run the commands in the *tex* section of verifiers.yml on my computer without failure.

A side note: The file generated by *pdflatex -halt-on-error list-nf.tex 2>&1 | tee ,latex-output* is **,latex-output**.

I am not sure whether this name is intended.  I'd rather expected **.latex-output**, for example, a hidden file not visible by default.

wlammen commented:
Now the cross check, using the develop branch:  I installed texlive-science, and after that ran the commands in the **tex** section of verifyers.yml again, in the develop environment.  As expected, the python script choked on the \\l sequence.  The _pdflatex -halt-on-error list-set.tex 2>&1 | tee ,latex-output_ command ends with failure:
...
> (/usr/share/texlive/texmf-dist/tex/latex/jknapltx/ursfs.fd)
! Undefined control sequence.
<recently read> \bm 
                    
> l.88 \verb!.+b! & $\bm
                      {+}$ \\
!  ==> Fatal error occurred, no output PDF file produced!
Transcript written on list-set.log.

So this pull request in total seems to improve on the situation.

Without any background in TEX or latex, for what it is worth, I approve this pull request.

jkingdon commented:
Fantastic work. Glad to hear that detecting more errors was as simple as checking for `no output PDF file produced` - I had imagined that we'd need to change more things than that but it is nice to be wrong on that sort of thing.

Checking iset.mm and nf.mm sounds good to me. I'm not aware of any reason we avoided them, we just didn't get around to adding them until now.

jkingdon commented:
> A side note: The file generated by _pdflatex -halt-on-error list-nf.tex 2>&1 | tee ,latex-output_ is **,latex-output**.
> 
> I am not sure whether this name is intended. I'd rather expected **.latex-output**, for example, a hidden file not visible by default.

I don't think it is an accident, if that is what you are wondering (a lot of the scripts in the `scripts` directory also put a comma at the start of names of log files and the like). I can speculate that having it start with a dot would have made it a bit more invisible than people preferred but to be honest I don't know much more than "we have always done it this way".
